### PR TITLE
fix(telemetry): Single project identifier

### LINKED ISSUE #507: Have a single way to identify a project in our telemetry
## Description

Let me show you a subset of the data that comes into our database from `kedro-telemetry` in our CLI commands table. 

| USER_ID          	| PROJECT_NAME    	| PACKAGE_NAME     	| USERNAME        	|
|------------------	|-----------------	|------------------	|-----------------	|
| 1343655360016970 	| 946346cad253d7b 	| 59f1c4c71309c46f 	| anonymous       	|
| 1856054624437891 	|                 	| 03e5c42f9fd8d0ab 	| ec3759e2c570d30 	|
| 8035489891863387 	|                 	| 8a171405cf81ef53 	| d92d8dd3de0a094 	|
| 4531280966080738 	|                 	| bb7f5d8666210fd5 	| ae46b8e451079db 	|
| 6022834494838396 	| 946346cad253d7b 	| 59f1c4c71309c46f 	| anonymous       	|
| 8035489891863387 	|                 	| 8a171405cf81ef53 	| d92d8dd3de0a094 	|
| 1416039240284711 	|                 	| 256cfde57fee6e63 	| 99adc231b045331 	|
| 6759436125356775 	| 2e35e5f9d892407 	| 2e35e5f9d8924072 	| anonymous       	|
| 6092965114078963 	|                 	| 0bc9045c71518347 	| b8308c21e8a66ad 	|
| 8249949466136913 	|                 	| ef85afadc4586f38 	| 57cbe0328faa502 	|

## Context

This task has two components: 

- Make a recommendation to the team about which is the most accurate identifier that we should use to ID a project
- And detail why `project_name` sometimes has blank entries 

## Expected Result
We should only have a single project ID that we all trust and we should not have a blank entry for that ID.

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
FIX #507 

Related docs update https://github.com/kedro-org/kedro/pull/3897

## Development notes
Manually tested the following scenario:

1. Created a project with consent
 - `project id` does not exist - it is generated and saved in `pyproject.toml`
 - `project id` exists - it is loaded from `pyproject.toml`, joined with the `package name`, hashed and used
 - package name changes,  `project id` exists - it is loaded from `pyproject.toml`, joined with the new `package name`, hashed and used

2. Created a project without consent

 - project id is not created

3. Packaging project

- package kedro project with modified `pyproject.toml` - works without errors
- run packaged project with consent=True and existing `pyproject.toml` - `project id` is generated and saved in `pyproject.toml`
- run packaged project with consent=True without `pyproject.toml` - `project id` is None

## Checklist

- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the relevant `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (noklam) [reply kedro-telemetry/kedro_telemetry/plugin.py:113]: https://docs.python.org/2/glossary.html#term-eafp

Another option, this happen in kedro-viz as well (I don't remember if we end up doing that). I find this more elegant if we don't care about the intermediate condition.

https://stackoverflow.com/questions/43491287/elegant-way-to-check-if-a-nested-key-exists-in-a-dict

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:118]: I think it's just the way to parse them into proper format before sending them to the Heap for logging: https://github.com/kedro-org/kedro-plugins/pull/557/files

So it's not affecting the way they're stored in `pyproject.toml`

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:88]: What if the user runs, say, `python -m spaceflights -p data_processing` (hence a packaged project) so the original `pyproject.toml` is not there, but they happen to have one in the current working directory?

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [inline kedro-telemetry/kedro_telemetry/plugin.py:253]: Does this mean that all packaged projects will have `project_uuid = "undefined_project_uuid"`? Isn't it better to send a `None` to Heap so we can outright ignore those?

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:88]: In that case, if consent is True and `pyproject.toml` is located in the same directory as `project_metadata. ' project_path` returns, it will be used to store the generated project UUID.

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:253]: We send `_hash(f"{project_uuid}{package_name}")` to the Heap. So we are still able to differentiate those projects by `package_name`. But if it's not the desired behaviour and we want to explicitly filter out projects for which UUID were not defined - then I can add that, so we send None instead.

Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:253]: Yeah, since we're salting `package_name` with `project_uuid`, in case the `project_uuid` is not found we should send `None` directly, to avoid accidentally using the same salt with 2 different package names

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:88]: Here's what I did:

1. I cloned a random Python project with a `pyproject.toml`, in this case https://github.com/mwaskom/seaborn/
2. While on it, I ran `python -m test_package_kedro --conf-source ../test-package-kedro/dist/conf-test_package_kedro.tar.gz`
3. The `pyproject.toml` file got modified:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 0a4e497d..72e91917 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,71 +1,34 @@
 [build-system]
-requires = ["flit_core >=3.2,<4"]
+requires = [ "flit_core >=3.2,<4",]
 build-backend = "flit_core.buildapi"
 
 [project]
 name = "seaborn"
 description = "Statistical data visualization"
-authors = [{name = "Michael Waskom", email = "mwaskom@gmail.com"}]
 readme = "README.md"
-license = {file = "LICENSE.md"}
-dynamic = ["version"]
-classifiers = [
-    "Intended Audience :: Science/Research",
-    "Programming Language :: Python :: 3.8",
-    "Programming Language :: Python :: 3.9",
-    "Programming Language :: Python :: 3.10",
-    "Programming Language :: Python :: 3.11",
-    "Programming Language :: Python :: 3.12",
-    "License :: OSI Approved :: BSD License",
-    "Topic :: Scientific/Engineering :: Visualization",
-    "Topic :: Multimedia :: Graphics",
-    "Operating System :: OS Independent",
-    "Framework :: Matplotlib",
-]
...
```

This behavior is introduced by this PR, it doesn't happen with `kedro-telemetry==0.4.0`.

I know it might seem like a special case, but I think writing to this file can have some unexpected consequences.

Some ideas that come to mind:

- We try to be more careful and constrain in which situations we write to the file
- We give up generating new UUIDs for existing projects, and instead we generate a random one on the fly when it's not available (giving the impression that there are more unique projects than there actually are)
  - Maybe this is related to the other comment
- ❓

Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: Thank you, @astrojuanlu for sharing this - I've tested the packaging a bit and found out one more problem. The package building fails when we have `project_uuid` under `[project]` section - so we should not store it there. I see the point of modifying other `pyproject.toml` - now we're breaking their formatting.

Based on the above I have two solutions in mind:

1. Store `project_uuid` in a separate custom section and append it to the end of the file, which should not change the formatting. Keep doing it for all the `pyproject.toml` as we do now
2. Store `project_uuid` in a separate custom section and append it to the end of the file. Differentiate whether it is kedro-related `pyproject.toml` or not by checking some keys such as `kedro_init_version` and modify only them.

Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: I thought https://pypi.org/p/toml/ had style-preserving write capabilities? (Or maybe not, see prior discussion at https://github.com/kedro-org/kedro/issues/3690#issue-2175592476)

Regardless, `project_uuid` should probably be stored under `[tools.kedro]`, the `[project]` table is only for PEP 621 keys https://peps.python.org/pep-0621/

Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: Yes, you're right - the only page one can customise is `[tool]`.

I've ended up with storing it under `[tools.kedro_telemetry]` which is appended to the end of `pyproject.toml`. That way we do not break the format of the original `pyproject.toml` cause we did it for kedro `pyproject.toml` as well. I didn't put it under `[tool.kedro]` since this approach requires `toml.dump` which actually changes formatting. I've tried some other encoders and it didn't help.

Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: I've tested the recent update with packaged project as well - and it works well now.

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:253]: Done

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py]: Done

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [inline kedro-telemetry/kedro_telemetry/plugin.py:260]: Nitpick: Shall we call this `project_id`? (since it's not going to be an UUID, but rather a hash of a combination of UUID + something else)

```suggestion
        "project_id": hashed_project_id,
```

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: Now, it's an expected behaviour—we do not change the formatting but append this to the end of the file. But we can always modify only the kedro-related  `pyproject.toml` but not the custom. The initial thought was that if people opt in for telemetry for packaged projects, they don't mind storing the `project_id` and `.telemetry` files. However, in reality, I think most people will opt out.

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [inline kedro-telemetry/kedro_telemetry/plugin.py]: Reading this logic again. The only case in which `_get_or_create_project_uuid(pyproject_path)` will return `None` is when the `pyproject.toml` file cannot be found, am I right?

**Review by astrojuanlu — COMMENTED** [review]
I didn't do a full code review but I think this is mostly in line with what's expected for #507.

However, I'm still observing a weird behavior for packaged projects, and since this PR introduces a change that can modify `pyproject.toml` files unrelated to Kedro, I opened https://github.com/kedro-org/kedro/issues/3902 in case it's something we want to look at more broadly. I'm fearing that this might lead to false positives (like: the user runs a packaged Kedro project, which does _not_ have a `project_uuid`, but since there happens to be a `pyproject.toml` file there, we get a false UUID). Would love to get opinions on that one before merging this PR.

One last thought: I'm assuming that we will introduce a change in `kedro new` so that projects have this `project_uuid` on `pyproject.toml` right after creation. Any objections/thoughts?

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py]: Yes, in all other cases, we either read it from there or generate and append it like so:
```
[tool.kedro_telemetry]
project_id = "..."
```

**Comment by ElenaKhaustova** [comment]
> I didn't do a full code review but I think this is mostly in line with what's expected for #507.
> 
> However, I'm still observing a weird behavior for packaged projects, and since this PR introduces a change that can modify `pyproject.toml` files unrelated to Kedro, I opened [kedro-org/kedro#3902](https://github.com/kedro-org/kedro/issues/3902) in case it's something we want to look at more broadly. I'm fearing that this might lead to false positives (like: the user runs a packaged Kedro project, which does _not_ have a `project_uuid`, but since there happens to be a `pyproject.toml` file there, we get a false UUID). Would love to get opinions on that one before merging this PR.
> 
> One last thought: I'm assuming that we will introduce a change in `kedro new` so that projects have this `project_uuid` on `pyproject.toml` right after creation. Any objections/thoughts?

I understand your arguments on modifying `pyproject.toml` files unrelated to Kedro. But I don't find it weird to add the `project_id` since we're asking a consent once again for packaged projects, meaning we're going to collect the metadata. The alternative is to store it out of `pyproject.toml` but previously we agreed we don't want to generate extra files and our goal is to collect metadata at the development stage rather than production, so we do not worry about propagating `project_id` generated at the development stage to the production. However, I don't mind differentiating `pyproject.toml` files unrelated to Kedro and do not modify them, in other words having `project_id=None `for them.

Theoretically we don't need `project_id` before sending the telemetry but generating it with `kedro new` can simplify it a bit for us. For example, we would be able to add it to the `[tool.kedro]` without rewriting the file (breaking the formatting) and then do not write to the `pyproject.toml` when collecting telemetry (which is better, IMO). But we would need to think about cases when a project is created without `kedro new` though I don't think they're common.

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (merelcht) [reply kedro-telemetry/kedro_telemetry/plugin.py:87]: > I'm still concerned about how packaged projects work but I'm going to open a separate issue about it.

@astrojuanlu are you concerned whether the project will still run or not or whether we collect telemetry correctly for packaged projects?

**Comment by merelcht** [comment]
I didn't immediately understand what @astrojuanlu meant with the `pyproject.toml` being changed with a packaged project but I managed to replicate that state and I agree with him that it's strange and probably not expected behaviour. However, one could argue that running a Kedro project inside a directory that contains another project and thus a `pyproject.toml` file is also unusual. 

That being said, I'd prefer it if we can ensure that the `pyproject.toml` file we're editing has a `[tool.kedro]` section. In fact, I'd argue that the `project_id` should be added to `[tool.kedro]` and not in an entirely new section for kedro-telemetry.

**Comment by ElenaKhaustova** [comment]
@merelcht, @astrojuanlu, thank you for sharing your thoughts. Based on them, I'll add logic to differentiate Kedro-related `pyproject.toml` and modify only it.

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (merelcht) [inline kedro-telemetry/kedro_telemetry/plugin.py:106]: Maybe we can make this a bit more specific by mentioning that the file doesn't contain a `tool.kedro` section?

**Review by merelcht — APPROVED** [review]
I left a minor comment, but other than that this now looks good to me and works as I expected 🙂

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:88]: Rather, the way the telemetry collector modifies its environment

**Review by astrojuanlu — APPROVED** [review]
Approved with 1 comment ⭐

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (ElenaKhaustova) [reply kedro-telemetry/kedro_telemetry/plugin.py:102]: I 100% agree, there are two reasons why I added it to the separate section for now. 

The first is because `toml.dump` changes formatting like so:
```
[project.optional-dependencies]
docs = [
    "docutils<0.18.0",
    "sphinx~=3.4.3",
    "sphinx_rtd_theme==0.5.1",
    "nbsphinx==0.8.1",
    "sphinx-autodoc-typehints==1.11.1",
    "sphinx_copybutton==0.3.1",
    "ipykernel>=5.3, <7.0",
    "Jinja2<3.1.0",
    "myst-parser~=0.17.2",
]
```

->

```
[project.optional-dependencies]
docs = [ "docutils<0.18.0", "sphinx~=3.4.3", "sphinx_rtd_theme==0.5.1", "nbsphinx==0.8.1", "sphinx-autodoc-typehints==1.11.1", "sphinx_copybutton==0.3.1", "ipykernel>=5.3, <7.0", "Jinja2<3.1.0", "myst-parser~=0.17.2",]
```
Which we do not like but I can replace it with custom file parsing to keep it proper.

2. The main one is because we validate expected keys in the `[tool.kedro]` section and raise an error in case of mismatch: https://github.com/kedro-org/kedro/blob/27f5405cefd6701ffac4c6243030486fb7d3c942/kedro/framework/startup.py#L107

So, to make it properly, we need changes on the framework side: possibly updating `kedro new`, as we discussed above. The question is whether we want to merge it like this for now or wait until we apply updates on the side of the framework.

@merelcht, @astrojuanlu

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [inline kedro-telemetry/kedro_telemetry/plugin.py:102]: One more thing (sorry for being a pain in the neck) can we properly catch the exception in case `pyproject.toml` isn't writeable for whatever reason? Thinking of read-only environments, permission issues, and the like

File: kedro-telemetry/kedro_telemetry/plugin.py
Comment (astrojuanlu) [reply kedro-telemetry/kedro_telemetry/plugin.py:102]: Ugh that "found unexpected keys" indeed would make any extra key in `[tool.kedro]` unfeasible for older Kedro versions. I don't have an opinion right now on whether it makes sense to allow more keys going forward, but it would not be a solution for our case due to lots of old Kedro projects out there in the wild, I think.

About style preservation, I'm wondering if tomlkit is a better choice here, but again this would introduce an extra dependency for very little gain.

Waiting for @merelcht's signoff but I'm okay moving forward with this.

Comment (merelcht) [reply kedro-telemetry/kedro_telemetry/plugin.py:102]: @ElenaKhaustova thanks for explaining! That's unfortunate, but in that case I'd also be in favour to just merge it like this. It's more of a "nice to have" than a must anyway.

**Pull request merged**
